src/EditorFeatures/TestUtilities/CodeActions/CodeFixVerifierHelper.cs

```diff
@@ -73,5 +87,150 @@ private static void VerifyMessageHelpLinkUri(DiagnosticAnalyzer analyzer)
                 Assert.NotEqual("", descriptor.HelpLinkUri ?? "");
             }
         }
+
+        public static Solution ApplyOptions(Solution solution, OptionsCollection options)
+        {
+            // For CodeStyle layer testing, we create an .editorconfig at project root
+            // to apply the options as workspace options are not available in CodeStyle layer.
+            // Otherwise, we apply the options directly to the workspace.
+
+#if CODE_STYLE
+            // We need to ensure that our projects/documents are rooted for
+            // execution from CodeStyle layer as we will be adding a rooted .editorconfig to each project
+            // to apply the options.
+            if (!options.IsEmpty)
+            {
+                solution = MakeProjectsAndDocumentsRooted(solution);
+                solution = AddAnalyzerConfigDocumentWithOptions(solution, options);
+            }
+#else
+            var optionSet = solution.Options;
+            foreach (var (key, value) in options)
+            {
+                optionSet = optionSet.WithChangedOption(key, value);
+            }
+
+            solution = solution.WithOptions(optionSet);
+#endif
+
+            return solution;
+        }
+
+#if CODE_STYLE
```

Comment:
like... this literally feels like entirely separate approaches here.  having us use ifdefs instead of other mechanisms (i.e. interfaces/DI/etc.) just feels really hacky and hard to maintain.